### PR TITLE
Add validation capability for user generated towncrier .rst files.

### LINKED ISSUE #9311: Add PR number and RST validation to towncrier workflow
Capturing the outcome from discussion at the Numba triage/public meeting on 2023-11-29.

In reviewing #9240 it transpired that it is possible to create a towncrier "news snippet" that has a filename that does not correspond to the PR against which the snippet applies. This results in strange problems where the release notes refer to references that don't exist. Also, following a number of reviews, it seems quite common for PR authors to make minor errors in the RST formatting when writing the snippet, this is not caught until towncrier attempts to render the snippet as release notes are created.

As a result, it was decided that the GHA workflow needs to run an RST validator against the snippet and in the command that runs the validator it should ensure that variable substitution takes place to make sure the PR number is in the name of the file to be validated.

### PR DESCRIPTION
**Pull request by kc611**
Fixes #9311,

As titled, 

This PR adds a validation script for user generated release notes / `.rst` file snippets for `towncrier` usage in their respective Pull requests. Also existing files have been passed through the validation and fixed accordingly.

### CODE REVIEW

File: maint/towncrier_rst_validator.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        encoding="utf-8",
+        stderr=STDOUT,
+    )
+
+    all_files = output.strip().splitlines()
+    rst_file = [file for file in all_files if 
+                file.startswith(rst_dir + "/" + str(args.pull_request_id) + ".")]
+    assert len(rst_file) == 1, "There must be a relevant .rst file added in the PR." + \
+        "Filename must start with the respective Pull Request ID" + \
+        " (eg 1234. for PR #1234) followed by a . character"
```
Comment (sklam) [inline maint/towncrier_rst_validator.py:38]: Suggestion:

- print content of `rst_file`
- Error message is confusing when no file is found. (I thought the script will let me run the validator on any correctly named .rst file but it turns out only modified files is checked. Btw, having a mode that can check any file even if they are not in the diff will be handy).

File: maint/towncrier_rst_validator.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    output = check_output(
+        ["git", "diff", "--name-only", "origin/main"],
+        encoding="utf-8",
+        stderr=STDOUT,
+    )
+
+    all_files = output.strip().splitlines()
+    rst_file = [file for file in all_files if 
+                file.startswith(rst_dir + "/" + str(args.pull_request_id) + ".")]
+    assert len(rst_file) == 1, "There must be a relevant .rst file added in the PR." + \
```
Comment (sklam) [inline maint/towncrier_rst_validator.py:36]:
```suggestion
    assert len(rst_file) == 1, "There must be a relevant .rst file added in the PR. " + \
```

**Review by sklam — COMMENTED** [review]
Next time, use `pathlib`. `pathlib.glob` can simplify things.

File: maint/towncrier_rst_validator.py (hunk 1 above)
Comment (kc611) [reply maint/towncrier_rst_validator.py]: Have added such a mode (and resolved above comments) in b42af38

File: maint/towncrier_rst_validator.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                    file.startswith(str(pr_id) + ".")]
+    else:
+        output = check_output(
+            ["git", "diff", "--name-only", "origin/main"],
+            encoding="utf-8",
+            stderr=STDOUT,
+        )
+        all_files = output.strip().splitlines()
+        rst_file = [file for file in all_files if 
+                    file.startswith(rst_dir + "/" + str(args.pull_request_id) + ".")]
```
Comment (sklam) [inline maint/towncrier_rst_validator.py:43]: RE: https://github.com/numba/numba/pull/9335#discussion_r1418113764

> print content of rst_file

I actually meant doing something like `print("found rst_files from git diff:", rst_file)` after this line.

For example, if you run: `python maint/towncrier_rst_validator.py --pull_request_id=9048` on this PR branch, you get:

```
Traceback (most recent call last):
  File "/Users/siu/dev/numba/maint/towncrier_rst_validator.py", line 50, in <module>
    file = list_rst_filename()
  File "/Users/siu/dev/numba/maint/towncrier_rst_validator.py", line 44, in list_rst_filename
    assert len(rst_file) == 1, f"No valid .rst file found in {rst_dir} for the given" + \
AssertionError: No valid .rst file found in docs/upcoming_changes for the given Pull Request ID: 9048. Valid .rst file should  have name starting with the respective Pull Request ID. (see numba/docs/upcoming_changes/README for more details)
```

I was confused that it did't recognize `./docs/upcoming_changes/9048.bug_fix.rst`. The actual problem was that the file is not part of the modified files; not that the filename is invalid.

**Review by sklam — COMMENTED** [review]
Thanks for adding the manual mode, it is handy for checking all the files with script like:

```python
In [9]: for f in files:
   ...:     f = os.path.basename(f)
   ...:     pr = f.split('.')[0]
   ...:     print(pr)
   ...:     if pr.isdigit():
   ...:         print(pr)
   ...:         prnum = int(pr)
   ...:         !python maint/towncrier_rst_validator.py --manual --pull_request_id $prnum
```

My only comment is below and sorry for my unclear comment in the previous review

File: maint/towncrier_rst_validator.py (hunk 3 above)
Comment (kc611) [reply maint/towncrier_rst_validator.py:50]: Added the suggested printing convention in [f158a19](https://github.com/numba/numba/pull/9335/commits/f158a197ae4f2f1b056b51ca8ce8b7b586707ed0)

**Review by sklam — APPROVED** [review]

**Pull request merged**